src/Starlight/Physics.hs

```diff
@@ -71,64 +73,71 @@ hit :: (Has (Reader (Seconds Float)) sig m, Has (Reader (System StateVectors)) s
      , Has (Reader (System StateVectors)) sig m
      , Has (State (System Body)) sig m
+     , HasCallStack
      )
   => CharacterIdentifier
   -> Character
   -> m Character
 runActions i c = do
   dt <- ask @(Seconds Float)
   system <- ask @(System StateVectors)
-  execState c (traverse_ (go dt system) (actions c)) where
-  go dt system = \case
-    Thrust -> do
-      rotation <- use (actor_.rotation_)
-      actor_ %= applyForce ((thrust ^*) <$> rotate rotation (unit _x)) dt
+  foldM (go dt system) c (actions c) where
+  go dt system c = \case
+    Thrust -> pure $ c & actor_ %~ applyForce ((thrust ^*) <$> rotate rotation (unit _x)) dt
 
-    Face dir -> do
-      actor <- use actor_
-      target <- fmap (either Body.actor Character.actor) <$> use (target_._Just.to (system !?))
-      for_ (direction actor target) $
-        modifying (actor_.rotation_) . face angular . fmap prj . angleOf . (^._xy) where
+    Face dir -> case direction (c^.actor_) target of
+      Just t  -> pure $! c & actor_.rotation_ %~ face angular (prj <$> angleOf (t^._xy))
+      Nothing -> pure c
+      where
       direction Actor{ velocity, position } t = case dir of
         Forwards  -> Just velocity
         Backwards -> t^?_Just.velocity_.to (subtract velocity) <|> Just (-velocity)
         Target    -> t^?_Just.to projected.to (`L.direction` position).coerced
 
-    Turn t -> actor_.rotation_ *= axisAngle (unit _z) (getRadians (case t of
+    Turn t -> pure $! c & actor_.rotation_ *~ axisAngle (unit _z) (getRadians (case t of
       L -> angular
       R -> -angular))
 
     Fire Main -> do
-      position <- use (actor_.position_)
-      rotation <- use (actor_.rotation_)
+      let position = c^.actor_.position_
       beams_ @Body %= (Beam{ colour = green, angle = snd (toAxisAngle rotation), position, firedBy = i }:)
+      pure c
 
-    ChangeTarget change -> target_ %= maybe (const Nothing) switchTarget change . (>>= (`elemIndex` identifiers)) where
+    ChangeTarget change -> pure $! c & target_ %~ maybe (const Nothing) switchTarget change . (>>= (`elemIndex` identifiers)) where
       elimChange prev next = \case { Prev -> prev ; Next -> next }
       switchTarget dir = \case
         Just i  -> identifiers !! i' <$ guard (inRange (0, pred (length identifiers)) i') where
           i' = elimChange (i - 1) (i + 1)  dir
         Nothing -> elimChange last head dir identifiers <$ guard (not (null identifiers))
       identifiers = System.identifiers system
+
+    Jump -> case target of
+      Just target
+        | distance (projected (c^.actor_)) (projected target) < 10000 -> pure c
+        | isFacing (pi/128) rotation targetAngle -> do
+          let distance' = distance (projected target) (projected (c^.actor_))
+          pure $! c & actor_.position_ +~ (1 - target^.magnitude_ / distance') *^ (projected target - projected (c^.actor_))
+        | otherwise                              -> go dt system c (Face Target) -- FIXME: face *near* the target
```

Comment:
I intend jumping to be more elegant/pretty—e.g. it should drop you into a stable orbit around planets—but this will do for now.